### PR TITLE
feat: add glob_match starlark utility function that support "*", "**" and "?" expressions as well as optionally allow for "*" and "?" to match of path separators

### PR DESCRIPTION
**Pull request by gregmagolan**
Default matching behaviour (with `match_path_separator = False`) follows https://metacpan.org/dist/File-Globstar/view/lib/File/Globstar.pod#*

### CODE REVIEW

File: lib/string_match.bzl [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    Args:
+        exp: the wildcard expression
+        str: the string against which to match the wildcard expression.
+
+    Returns:
+        True if the string matches the wildcard expression
+    """
+
+    # Reduce all duplicate wildcards to single wildcards
```
Comment (jbedard) [inline lib/string_match.bzl:25]: This is unrelated to glob `**` right? It just simplifies handling of double-wildcards which are pointless?

File: lib/string_match.bzl [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        else:
+            break
+
+    if exp.find("?") != -1:
+        fail("string_match: `?` expression not supported")
+
+    exp_i = 0
+    str_i = 0
+
+    # HACK: since "?" will never be in the expression, use it to
```
Comment (jbedard) [inline lib/string_match.bzl:39]: But it soon will be! 🤷

Comment (gregmagolan) [reply lib/string_match.bzl:39]: :shrug:

The principled way is to split on `*` and inject a `*` in between in element but not on the ends

File: lib/string_match.bzl (hunk 1 above)
Comment (gregmagolan) [reply lib/string_match.bzl:25]: Yeah, if you search for any sequence of characters twice you're not doing anything useful the 2nd time. Infinity * 2 is still infinity.

File: lib/string_match.bzl [hunk 3]
```diff
@@ -0,0 +1,88 @@
+"""
+Basic string match implementation with wildcards for starlark.
+
+This was originally developed by @jbedard for use in rules_js (https://github.com/aspect-build/rules_js)
```
Comment (jbedard) [inline lib/string_match.bzl:4]: Worth linking to the commit in the repo instead of just repo? 🤷

**Review by jbedard — APPROVED** [review]

File: lib/string_match.bzl (hunk 1 above)
Comment (gregmagolan) [reply lib/string_match.bzl]: No longer applies with latest changes

File: lib/string_match.bzl (hunk 3 above)
Comment (gregmagolan) [reply lib/string_match.bzl]: Good idea

File: lib/string_match.bzl (hunk 2 above)
Comment (gregmagolan) [reply lib/string_match.bzl]: Fixed

File: lib/string_match.bzl (hunk 3 above)
Comment (gregmagolan) [reply lib/string_match.bzl]: ✅
